### PR TITLE
Fix migration that introduces instance states.

### PR DESCRIPTION
**Pull request by itsjeyd**
Follows up on #57.

With the previous version of the migration, instance states were getting updated multiple times and ended up being set to `Terminated` if an instance had one or more servers with status `Terminated` associated with them.

Actual results when running migrations were:

```
Instance 66: Running, Server: Ready
Instance 71: Terminated, Server: Ready
Instance 79: Running, Server: Ready
Instance 90: Terminated, Server: Ready
Instance 104: Running, Server: Ready
Instance 126: Terminated, Server: Ready
Instance 127: Terminated, Server: Ready
Instance 146: Configuration failed, Server: Ready
Instance 170: Terminated, Server: Ready
Instance 172: Terminated, Server: Ready
Instance 176: Terminated, Server: Ready
Instance 177: Terminated, Server: Ready
Instance 178: Configuration failed, Server: Ready
Instance 179: Terminated, Server: Ready
Instance 181: Terminated, Server: Ready
Instance 182: Terminated, Server: Ready
Instance 183: Terminated, Server: Ready
Instance 186: Configuration failed, Server: Ready
Instance 187: Running, Server: Ready
Instance 188: Terminated, Server: Ready
Instance 189: Configuration failed, Server: Ready
Instance 190: Terminated, Server: Ready
Instance 191: Terminated, Server: Ready
Instance 192: Terminated, Server: Ready
Instance 193: Terminated, Server: Ready
Instance 196: Terminated, Server: Ready
Instance 197: Terminated, Server: Ready
```

Expected results:

```
Instance  66: Running, Server: Ready
Instance  71: Running, Server: Ready
Instance  79: Running, Server: Ready
Instance  90: Running, Server: Ready
Instance 104: Running, Server: Ready
Instance 126: ConfigurationFailed, Server: Ready
Instance 127: Running, Server: Ready
Instance 146: ConfigurationFailed, Server: Ready
Instance 170: Running, Server: Ready
Instance 172: Running, Server: Ready
Instance 176: Running, Server: Ready
Instance 177: Running, Server: Ready
Instance 178: ConfigurationFailed, Server: Ready
Instance 179: Running, Server: Ready
Instance 181: Running, Server: Ready
Instance 182: ConfigurationFailed, Server: Ready
Instance 183: Running, Server: Ready
Instance 186: ConfigurationFailed, Server: Ready
Instance 187: Running, Server: Ready
Instance 188: Running, Server: Ready
Instance 189: ConfigurationFailed, Server: Ready
Instance 190: Running, Server: Ready
Instance 191: ConfigurationFailed, Server: Ready
Instance 192: ConfigurationFailed, Server: Ready
Instance 193: ConfigurationFailed, Server: Ready
Instance 196: Running, Server: Ready
Instance 197: Running, Server: Ready
```

@open-craft/core Please do not deploy commits that are newer than 5319f4e23caf775619a281f857e09760a851ddee before this gets merged.

### CODE REVIEW

**Comment by itsjeyd** [comment]
@bradenmacdonald Please review.

File: instance/migrations/0042_add_instance_status.py [hunk 1]
```diff
@@ -6,40 +6,37 @@
...
+        return InstanceStatus.Running.state_id
+    if server_status == "provisioning":
+        if server_progress in ("running", "success"):
+            return InstanceStatus.ConfiguringServer.state_id
+        elif server_progress == "failed":
+            return InstanceStatus.ConfigurationFailed.state_id
+    if server_status == "rebooting":
+        return InstanceStatus.ConfiguringServer.state_id
+    if server_status == "terminated":
+        return InstanceStatus.Terminated.state_id
```
Comment (bradenmacdonald) [inline instance/migrations/0042_add_instance_status.py]: Just a note that didn't occur to me before: It's better for migrations to not have any reference to the current state of the app, but we have some coupling here with `from instance.models.instance import Status as InstanceStatus`.

If we ever rename or remove one of these status types in the future, we'd need to modify this migration. So even though it feels wrong, it's usually better to just put string constants (`"waiting"` vs. `InstanceStatus.WaitingForServer.state_id`) in migrations.

I don't think it's worth making that change now though; we can easily cross that bridge when we come to it, if it ever even happens which is unlikely.

**Comment by bradenmacdonald** [comment]
Code looks good and this change makes sense to me. Going to try some local testing now.

File: instance/migrations/0042_add_instance_status.py [hunk 2]
```diff
@@ -6,40 +6,37 @@
...
-        server_set___status="rebooting").update(
-            _status=InstanceStatus.ConfiguringServer.state_id
-        )
-    SingleVMOpenEdXInstance.objects.filter(
-        server_set___status="terminated").update(
-            _status=InstanceStatus.Terminated.state_id
-        )
+    for instance in SingleVMOpenEdXInstance.objects.iterator():
+        current_server = get_current_server(instance)
+        server_status = current_server._status
```
Comment (bradenmacdonald) [inline instance/migrations/0042_add_instance_status.py]: Getting an error here when there is no server attached to an instance (which happens when there are more instances created than workers are available to spawn servers):

```
  Applying instance.0042_add_instance_status...Traceback (most recent call last):
  File "./manage.py", line 37, in <module>
    execute_from_command_line(sys.argv)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/__init__.py", line 351, in execute_from_command_line
    utility.execute()
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/__init__.py", line 343, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/commands/migrate.py", line 222, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/db/migrations/executor.py", line 110, in migrate
    self.apply_migration(states[migration], migration, fake=fake, fake_initial=fake_initial)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/db/migrations/executor.py", line 148, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/db/migrations/migration.py", line 115, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/db/migrations/operations/special.py", line 183, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/vagrant/instance/migrations/0042_add_instance_status.py", line 36, in set_instance_states
    server_status = current_server._status
AttributeError: 'NoneType' object has no attribute '_status'
```

Just needs an `if current_server:` and to handle setting the status to `New` in that case.

**Comment by bradenmacdonald** [comment]
👍  if tests pass.

**Comment by bradenmacdonald** [comment]
@itsjeyd: Do you know what happened with the build here?

```
29/Apr/2016 20:52:59 |     INFO | process=11764 | instance.models.instance  | instance=e006c08e.integration | Provisioning completed
29/Apr/2016 20:52:59 |     INFO | process=11764 | instance.models.instance  | instance=e006c08e.integration | Rebooting server 9ca544c7-8313-4177-ae2c-a1f2b3a28017...
29/Apr/2016 20:52:59 |     INFO | process=11764 | requests.packages.urllib3 | Resetting dropped connection: compute.gra1.cloud.ovh.net
29/Apr/2016 20:53:00 |     INFO | process=11764 | requests.packages.urllib3 | Resetting dropped connection: compute.gra1.cloud.ovh.net
29/Apr/2016 20:53:31 |     INFO | process=11764 | instance.models.server    | instance=e006c08e.integr,server=9ca544c7 | Waiting to reach status from which we can proceed...
29/Apr/2016 20:53:31 |     INFO | process=11764 | requests.packages.urllib3 | Resetting dropped connection: compute.gra1.cloud.ovh.net
.... repeated several times, then it times out
```

**Comment by smarnach** [comment]
@bradenmacdonald We are frequently getting the "Resetting dropped connection: compute.gra1.cloud.ovh.net" messages, but usually the build succeeds anyway.  I don't know what is causing them, but I concluded that it must be harmless.

**Comment by bradenmacdonald** [comment]
@smarnach But in this case, the build printed "Waiting to reach status from which we can proceed..." and then nothing but those "Resetting dropped connection" messages in clusters of 3 every 2-3 minutes for 57 minutes until the build timed out.

Earlier in the same build and in my own experience testing this, I have seen those same messages, but usually they stop after less than a minute and the build/provisioning proceeds.

**Comment by itsjeyd** [comment]
@bradenmacdonald @smarnach I'm not exactly sure what happened but I think CircleCI was having issues: [This build](https://circleci.com/gh/open-craft/opencraft/657) failed with "bin/run-circleci-tests died unexpectedly" and has a message at the top that says:

> Looks like we had a bug in our infrastructure, or that of our providers (generally GitHub or AWS) We should have automatically retried this build. We've been alerted of the issue and are almost certainly looking into it, please contact us if you're interested in the cause of the problem.

The fact that the [next build](https://circleci.com/gh/open-craft/opencraft/658) timed out might be related to this issue. CircleCI automatically ran [another build](https://circleci.com/gh/open-craft/opencraft/659) after that one, and it finished successfully.

**Pull request merged**